### ISSUE TITLE
request for str subcommands

### ISSUE THREAD

**Issue by tiawl** [modernish/modernish#20]
Hi

I would like to request a new feature. For this simple program I am a little bit annoyed:
```
#! /usr/bin/env modernish
#! use safe
#! use var/arith

print_only_N_first_chars ()
{
  if gt ${1} 0 && str isint ${1}
  then
    command printf %.${1}s\\n ${2}
  fi
}

print_only_N_first_chars 10 AN-AMAZING-STRING
print_only_N_first_chars 0x0A AN-AMAZING-STRING
```

The first call is an expected use case for this function. The next call however is not: `0x0A` is a valid integer so it passes the `str isint` test to finally crash the `printf` call because `printf` expects a positive decimal integer.

What do you think about adding this set of subcommands to the `str` function:
- `str isdec`: test if the string is a valid decimal
- `str ishex`: test if the string is a valid hexadecimal
- `str isoct`: test if the string is a valid octal

**Comment by McDutchie** [comment]
That goes uncomfortably in the direction of featuritis, I'm afraid. `str isint` checks that the string is a valid integer number for the purposes of arithmetic expressions. An `$((`arithmetic expansion`))` can be used to convert any arithmetic expression to a decimal number, so there is already an easy way to make your script work; see below.

Also, `str isint` should be called before `gt` as the latter will error out on an invalid arithmetic expression.

```sh
#! /usr/bin/env modernish
#! use safe
#! use var/arith

print_only_N_first_chars ()
{
  if str isint ${1} && gt ${1} 0
  then
    command printf %.$(($1))s\\n ${2}
  fi
}

print_only_N_first_chars 10 AN-AMAZING-STRING
print_only_N_first_chars 0x0A AN-AMAZING-STRING
```

**Comment by tiawl** [comment]
> `str isint` checks that the string is a valid integer number for the purposes of arithmetic expressions.

I did not understand the purpose of `str isint` like this. I thought that can be used to check an integer before any command which needs this number. Not specificly for arithmetic expressions. It is more clear now.

> An `$((` arithmetic expansion `))` can be used to convert any arithmetic expression to a decimal number

Nice tip, I did not know this one, thank you.

> That goes uncomfortably in the direction of featuritis

I did not know the term of featuritis. I read about this with a lot of interest. However, I think my request:
1. is not totally outside of the `modernish`'s design.
2. could be profitable for other users,

First of all, because I did not know that:
> An `$((` arithmetic expansion `))` can be used to convert any arithmetic expression to a decimal number

and because my last small program was written without checking, my request is lacking of seriousness.

I agree with you, checking if a string is a decimal integer is pointless. I am not going to talk more about `str isdec`.
If another tip like that exists for hexadecimal integer or octal integer numbers, without using subshell or external command, there are no reason to talk about `str ishex`/`str isoct` so I will admit that I am wrong and I will close this issue.
However, because POSIX defines tools like `bc`, `dd`, `od` or `iconv` (and maybe other that I do not know), I think that checking if the input is correct before using it with these utilities is not superfluous.
And this is specificly when a user is playing with those utilities that he is also dealing with octal and hexadecimal integers.
So yes, before each of my script, when I am using those utilities, I can write 2 small functions with a `case` statement into each of them to allow me to check for hexadecimal or octal integers.
But that is not the place where I want to see this function. For me if I have to write those functions before each of my `bc`/`od`/`dd`/`iconv` scripts, the place of this function is in a library.
